exporters/metric/dynamicconfig/servicereader.go

```diff
@@ -0,0 +1,95 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dynamicconfig
+
+import (
+	"context"
+	"time"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+
+	resourcepb "github.com/open-telemetry/opentelemetry-proto/gen/go/resource/v1"
+	pb "github.com/vmingchen/opentelemetry-proto/gen/go/collector/dynamicconfig/v1"
+)
+
+type ServiceReader struct {
+	// Required
+	configHost string
+
+	// Timestamp of last time config service was checked
+	lastTimestamp time.Time
+
+	// Most recent config version
+	lastKnownFingerprint []byte
+
+	// Suggested time to wait before checking config service again (seconds)
+	suggestedWaitTimeSec int32
+
+	// Required. Label to identify this instance
+	resource *resourcepb.Resource
+}
+
+func newServiceReader(configHost string, resource *resourcepb.Resource) *ServiceReader {
+	return &ServiceReader{
+		configHost: configHost,
+		resource:   resource,
+	}
+}
+
+// Reads from a config service
+func (r *ServiceReader) readConfig() (*Config, error) {
+	// Wait for the suggestedWaitTimeSec
+	if !r.lastTimestamp.IsZero() && r.suggestedWaitTimeSec != 0 {
+		suggestedReadTime := r.lastTimestamp.Add(time.Duration(r.suggestedWaitTimeSec) * time.Second)
+		time.Sleep(suggestedReadTime.Sub(time.Now()))
```

Comment:
How does this work with the checkChanges() in notifier? There is a tick over there, and we have a sleep(). Are they adding up to the suggested read time?

Reply:
Yes. So `suggestedReadTime` is the earliest time we are suggested to read from the config service again.

Say we read from the service at 5:00am, with a `suggestedReadTimeSec` of 300, then our `suggestedReadTime` is 5:05am.

If we get to this part of the code at 5:03am, we will sleep for 2 more minutes.

The reason I did it this way is because if we did it in the notifier, this scenario is possible:
-We read from the config service every 10 minutes.
-At 5:00am, we read a config with a suggestedWaitTimeSec of 11*60, or 11 minutes.
-At 5:10am, we haven't waited suggestedWaitTimeSec, so we wait for another `tick`.
-At 5:20am, we read a config again finally. 

This way I think we wait a more accurate amount of time.